Add celltyping to external instructions (#502)
AlexsLemonade/scpca-nf: external-instructions.md

    - [Additional reference files](#additional-reference-files)
    - [Downloading container images](#downloading-container-images)
- [Repeating mapping steps](#repeating-mapping-steps)

> - [Cell type annotation](#cell-type-annotation)

Reviewer: Let's add a TOC title?

  - [Libraries with additional feature data (ADT or cellhash)](#libraries-with-additional-feature-data-adt-or-cellhash)
  - [Multiplexed (cellhash) libraries](#multiplexed-cellhash-libraries)
  - [Spatial transcriptomics libraries](#spatial-transcriptomics-libraries)
- [Cell type annotation](#cell-type-annotation)
  - [Providing existing cell type labels](#providing-existing-cell-type-labels)
  - [Performing cell type annotation](#performing-cell-type-annotation)
    - [Specifying cell type annotation reference files](#specifying-cell-type-annotation-reference-files)
  - [Repeating cell type annotation](#repeating-cell-type-annotation)
- [Output files](#output-files)

> <!-- END doctoc generated TOC please keep comment here to allow auto update -->

Author: yes. yes we do 🌮

| `sample_reference`       | The name of the reference to use for mapping, available references include: `Homo_sapiens.GRCh38.104` and `Mus_musculus.GRCm39.104`                                                                                                                                                                                                                                                                                                                         |
| `files_directory`        | The full path/uri to directory containing fastq files (unique per run)                                                                                                                                                                                                                                                                                                                                                                                      |

The following optional columns may be necessary for running other data modalities (CITE-seq, spatial transcriptomics) or including existing cell type labels:

> | column_id                   | contents                                                                                                                                                                                                                     |
> | --------------------------- | ---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |

Reviewer: Because I agree that wide table formatting is annoying.
```suggestion
<!-- prettier-ignore -->
| column_id | contents |
| --------- | -------- |
```

Feel free to remove all the extra spaces below as well!

   You will need to provide the path/uri to this file as a workflow parameter, which you will need to define in your configuration file.
   For more information on adding parameters to your configuration file, see [Configuring scpca-nf for your environment](#configuring-scpca-nf-for-your-environment).
3. Run the workflow with the `--perform_celltyping` flag.

Once you have followed the above steps and added the path/uri to the `celltype_project_metafile` to your configuration file, you can use the following command to run the workflow with cell type annotation:

```sh
nextflow run AlexsLemonade/scpca-nf \
  --perform_celltyping
```

### Choosing reference datasets

The Data Lab has compiled several references, listed in [`celltype-reference-metadata.tsv`](references/celltype-reference-metadata.tsv).
All references listed in this table are publicly available on S3 for use with cell type annotation.
It is possible to provide your own references as well; instructions for this are forthcoming.
Note that you must use one of the references described here to be eligible for inclusion in the ScPCA Portal.

#### `SingleR` references

> By default, `SingleR` annotation uses references from the [`celldex` package](https://bioconductor.org/packages/release/data/experiment/html/celldex.html).

Reviewer 2: Can you file an issue to update this section on cell type references, after #565 goes in.

Author: fyi im on this, have a note waiting for it :)

  --repeat_mapping
```

## Cell type annotation

`scpca-nf` can perform cell type annotation using two complementary methods: the reference-based method [`SingleR`](https://bioconductor.org/packages/release/bioc/html/SingleR.html) and the marker-gene based method [`CellAssign`](https://github.com/Irrationone/cellassign).

By default, no cell type annotation is performed.
You can turn on cell type annotation by taking the following steps:

1. Select appropriate reference dataset(s) to use with each method of interest.
2. Prepare a `celltype_project_metafile` TSV (described below) to provide reference dataset information for each of `SingleR` and `CellAssign` to the workflow.
   You will need to provide the path/uri to this file as a workflow parameter, which you can specify either on the command line or define in your configuration file.
3. Run the workflow with the `--perform_celltyping` flag.

> You would then run the workflow with cell type annotation from the command line as, for example,

Reviewer: Making phrasing more consistent with previous code chunks? (Notably the colon)
```suggestion
You would then run the workflow with cell type annotation with a  command line such as the following:
```

> You would then run the workflow with cell type annotation from the command line as, for example,
> 
> ```sh
> nextflow run AlexsLemonade/scpca-nf \
>   --perform_celltyping \
>   --celltype_project_metafile examples/example_project_celltype_metadata.tsv
> ```

Reviewer 2: ```suggestion
Once you have followed the above steps and added the path/uri to the `celltype_project_metafile` to your configuration file, you can use the following command to run the workflow with cell type annotation:

```sh
nextflow run AlexsLemonade/scpca-nf \
  --perform_celltyping 
```
```
I want to be explicit that the above steps need to be completed. Also I'm removing defining the path here, but noting that it should be in the config file.

### Choosing reference datasets

> The Data Lab has compiled several references, listed in [`celltype-reference-metadata.tsv`](references/celltype-reference-metadata.tsv) and available from S3, for use in cell type annotation.

Reviewer 2: ```suggestion
The Data Lab has compiled several references, listed in [`celltype-reference-metadata.tsv`](references/celltype-reference-metadata.tsv). 
All references listed in this table are publicly available on S3 for use with cell type annotation.
```

It is possible to provide your own references as well; instructions for this are forthcoming.
Note that you must use one of the references described here to be eligible for inclusion in the ScPCA Portal.

#### `SingleR` references

By default, `SingleR` annotation uses references from the [`celldex` package](https://bioconductor.org/packages/release/data/experiment/html/celldex.html).

> Available references include `BlueprintEncodeData`, `DatabaseImmuneCellExpressionData`, `HumanPrimaryCellAtlasData`, and `MonacoImmuneData`.

Reviewer 2: We currently don't support these, and don't plan to. But should we just add them when we build references, even if we don't use them right away?

Reviewer 2: You are correct! This is why we need a list somewhere other than S3 😂

Author: > This is why we need a list somewhere other than S3 😂

☝️☝️☝️☝️☝️

Please consult the [`celldex` documentation](https://bioconductor.org/packages/release/data/experiment/vignettes/celldex/inst/doc/userguide.html) to determine which of these references, if any, is most suitable for your dataset.

> `SingleR` reference files that you can specify in the `celltype_project_metafile` are stored in `s3:://scpca-references/celltype/singler_models` and are formatted as `<singler_ref_name>_celldex_<celldex_version>_model.rds`.

Reviewer 2: A thought I had reading this... do we want to provide a list of currently available ref files and filenames in the repo for users to easily find which reference would be suitable for them? I'm thinking we could write out this file as part of `build-celltype-ref.nf`.

Reviewer: If we want to write out such a file, we would probably want to keep it on S3 still? A JSON file with all of the currently supported references would be pretty nice, and it could actually simplify the code in #565 a bit.

Reviewer 2: Is there any reason we couldn't write it out to the repo? And store in `references`? We can provide the output directory and use `${projectDir}` to tell Nextflow where to store it?

Reviewer 2: We could take it one step further and create a GHA for running the reference workflow anytime the metadata or workflow gets updated... then that would ensure that file stays up to date with changes. But first we should get the file to be written out.

Author: > A JSON file with all of the currently supported references would be pretty nice, and it could actually simplify the code in https://github.com/AlexsLemonade/scpca-nf/pull/565 a bit.

So far this seems the like move to me, written out directly to S3 by the references workflow. 
I wonder if we could work out the opposite of our `sample_info` S3 syncing; when the JSON file changes in S3, file a PR to `scpca-nf`? I don't exactly know how we'd set that up, though. Maybe something with time stamps..?

For example, consider the reference file `BlueprintEncodeData_celldex_1-10-1_model.rds`:

- The reference name is `BlueprintEncodeData`.
- The `celldex` version is `1-10-1`.

#### `CellAssign` references

By default, `CellAssign` annotation uses marker gene set references from [PanglaoDB](https://panglaodb.se/), as compiled by the Data Lab to represent common organ/tissue groupings.

> Available references include `blood-compartment`, `brain-compartment`, and `muscle-compartment`.

Reviewer 2: We're going to have more available combinations, so I don't like explicitly stating that here. I think you could state these are examples, but I think we should compile a list of all our supported options and the file names for users.

The specific organs used to compile marker gene references are listed in [`celltype-reference-metadata.tsv`](references/celltype-reference-metadata.tsv).

`CellAssign` marker gene reference files that you can specify in the `celltype_project_metafile` are stored in `s3:://scpca-references/celltype/cellassign_references` and are formatted as `<cellassign_ref_name>_PanglaoDB_<date>.tsv`.
For example, consider the reference file `blood-compartment_PanglaoDB_2020-03-27.tsv`:

- The reference name is `blood-compartment`.
- The `PanglaoDB` date (which represents the `PanglaoDB` version) is `2020-03-27`.

### Preparing the cell type project metadata file

All libraries within a given project will use the same reference dataset for each of `SingleR` and `CellAssign`, respectively.

> The `celltype_project_metafile` file should contain these five columns with the following information (see the example file in [`examples/example_project_celltype_metadata.tsv`](examples/example_project_celltype_metadata.tsv)):

Reviewer 2: Can you turn the link to the example project file into a button? Like we have the other example links.

| column_id             | contents                                                                                                                              |
| --------------------- | ------------------------------------------------------------------------------------------------------------------------------------- |

> | `scpca_project_id`    | Project ID matching values in the metadata file                                                                                       |

Reviewer 2: I would link to the section for preparing the metadata file here.

| `singler_ref_name`    | Reference name for `SingleR` annotation, e.g., `BlueprintEncodeData`. Use `NA` to skip `SingleR` annotation                           |
| `singler_ref_file`    | Path/uri to `SingleR` reference file, e.g., `BlueprintEncodeData_celldex_1-10-1_model.rds`. Use `NA` to skip `SingleR` annotation     |
| `cellassign_ref_name` | Reference name for `CellAssign` annotation, e.g. `blood-compartment`. Use `NA` to skip `CellAssign` annotation                        |
| `cellassign_ref_file` | Path/uri to `CellAssign` reference file, e.g., `blood-compartment_PanglaoDB_2020-03-27.tsv`. Use `NA` to skip `CellAssign` annotation |

### Repeating cell type annotation

> When cell typing is turned on with `--perform_celltyping`, `scpca-nf` will skip annotation for any libraries whose cell type annotation results already exist in the `checkpoints` folder of the given library's output directory, and those reference dataset version is unchanged.

Reviewer 2: Just a note that we are actually checking the files and not the version.

> This saves substantial processing time and cost if the cell type annotation reference versions are unchanged.

Reviewer: (people's perceived costs may not depend on time if they are running on a "free" server)
```suggestion
This saves substantial processing time if the cell type annotation reference versions are unchanged.
```

However, you may wish to repeat the cell typing process if there have been other changes to the data.

To force repeating the cell type annotation process, use the `--repeat_celltyping` flag along with the `--perform_celltyping` flag at the command line:

```sh

> nextflow run AlexsLemonade/scpca-nf \
>   --perform_celltyping \
>   --repeat_celltyping \
>   --celltype_project_metafile examples/example_project_celltype_metadata.tsv

Reviewer 2: ```suggestion
nextflow run AlexsLemonade/scpca-nf \
  --perform_celltyping \
  --repeat_celltyping 
```

```

> ### Providing existing cell type labels

Reviewer 2: Just noting that I think this is actually sufficient or providing submitter cell types, so you can probably ignore my earlier comment about temporarily removing it. I do think we want to add an example file though.

Author: I realize I misinterpreted your earlier comment! When you said move this to later on, you meant do it later on (#440) but I read it as "move it later in the document"! So....maybe it worked out ok? example file - https://github.com/AlexsLemonade/scpca-nf/pull/502/commits/38dba37ba7d0321d4c046dc0d01bb7e7b3b4f9cb

If not, I can remove all this and stash to tackle it later.

Reviewer 2: No I did mean to move it down here, but also that we could tackle it separately. I think what you have now is fine though!

_Note: The workflow is currently set up to work only with spatial transcriptomic libraries produced from the [Visium Spatial Gene Expression protocol](https://www.10xgenomics.com/products/spatial-gene-expression) and has not been tested using output from other spatial transcriptomics methods._

## Cell type annotation

### Providing existing cell type labels

If you have already performed cell type annotation and wish to include these labels in the final workflow results, you can include the column `submitter_cell_types_file` in your run metadata file ([see example here](examples/example_run_metadata.tsv)).
This column should be filled with the path or uri to a TSV file with existing cell type labels.

This file _must_ include the following columns:

| column_id              | contents                                        |
| ---------------------- | ----------------------------------------------- |
| `scpca_library_id`     | Library ID matching values in the metadata file |
| `cell_barcode`         | The cell id with the given annotation label     |
| `cell_type_assignment` | The annotation label for that cell              |

> Optionally, you can also include a column `cell_type_ontology` with ontology labels corresponding to the given annotation label.

Author: >  then I would say, but wait, if you want to provide your own labels you can do that in addition to or instead.

Fighting the urge to write "But wait, there's more!"

### Performing cell type annotation

`scpca-nf` can perform cell type annotation using two complementary methods: the reference-based method [`SingleR`](https://bioconductor.org/packages/release/bioc/html/SingleR.html) and the marker-gene based method [`CellAssign`](https://github.com/Irrationone/cellassign).

> You can turn on cell type annotation by using the `--perform_celltyping` flag.

Author: Noting that given other broad rearrangements, I kept the references section below the the explanation of this metadata file. Let's see how it looks now in the context of other changes, and can still move it if it needs to move!

> You can turn on cell type annotation by using the `--perform_celltyping` flag.

Author: Just kidding!! References now above the metadata description - https://github.com/AlexsLemonade/scpca-nf/pull/502/commits/5e8aafd68e1106e1b2681770ac3fd02eb4bf8eba

After building the docker image, you will need to push it to a [private docker registry](https://www.docker.com/blog/how-to-use-your-own-registry/) and set `params.SPACERANGER_CONTAINER` to the registry location and image id in the `user_template.config` file.
*Note: The workflow is currently set up to work only with spatial transcriptomic libraries produced from the [Visium Spatial Gene Expression protocol](https://www.10xgenomics.com/products/spatial-gene-expression) and has not been tested using output from other spatial transcriptomics methods.*

> ## Cell type annotation

Reviewer: I think this whole section probably should be moved above the special considerations section. I might actually move Output files up as well, as I think people probably want to know about outputs for the normal workflow before details for other cases, but harder to say there.

| `cellassign_ref_name` | Reference name for `CellAssign` annotation. Must be one of `blood-compartment`, `brain-compartment`, or `muscle-compartment`. Use `NA` to skip `CellAssign` annotation |
| `cellassign_ref_file` | Path to internal `CellAssign` reference file. The expected format for this file is described below. Use `NA` to skip `CellAssign` annotation |

This file can be specified at the command line (shown below) or defined in your configuration file.
For example, you would run from the command line as:

```sh
nextflow run AlexsLemonade/scpca-nf \
  --perform_celltyping \
  --celltype_project_metafile examples/example_project_celltype_metadata.tsv
```

#### Specifying cell type annotation reference files

The `singler_ref_file` and `cellassign_ref_file` names follow a specific format corresponding to available references in S3.

- `SingleR` annotation uses references from the [`celldex` package](https://bioconductor.org/packages/release/data/experiment/html/celldex.html).
Available reference options include `BlueprintEncodeData`, `DatabaseImmuneCellExpressionData`, `HumanPrimaryCellAtlasData`, and `MonacoImmuneData`.

>   - Available `SingleR` reference files that you can specify in the `celltype_project_metafile` are saved in `s3:://scpca-references/celltype/singler_models` and are formatted as `<singler_ref_name>_celldex_<celldex_version>_model.rds`.

Author: > I'd just name the default here

Do you mean the BlueprintEncode file name?

  - Please consult the [`celldex` documentation](https://bioconductor.org/packages/release/data/experiment/vignettes/celldex/inst/doc/userguide.html) to determine which of these references, if any, is most suitable for your dataset.
- `CellAssign` annotation uses marker gene set references from [PanglaoDB](https://panglaodb.se/).
  Organ-based references were compiled by the Data Lab to represent common organ/tissue groupings.

>   Available references include `blood-compartment`, `brain-compartment`, and `muscle-compartment`.

Author: @allyhawkins where are the specific organs relevant to each reference documented? And/or where is the code that creates these?

I looked in `references` but then just ended up opening this issue #561.. I don't immediately see anything in `scpca-admin` either?

>   Available references include `blood-compartment`, `brain-compartment`, and `muscle-compartment`.

Reviewer: This file should be referenced, as it has the main details required: https://github.com/AlexsLemonade/scpca-nf/blob/main/references/celltype-reference-metadata.tsv

As an example, the Dockerfile that we used to build Space Ranger can be found [here](https://github.com/AlexsLemonade/alsf-scpca/tree/main/images/spaceranger).

After building the docker image, you will need to push it to a [private docker registry](https://www.docker.com/blog/how-to-use-your-own-registry/) and set `params.SPACERANGER_CONTAINER` to the registry location and image id in the `user_template.config` file.
_Note: The workflow is currently set up to work only with spatial transcriptomic libraries produced from the [Visium Spatial Gene Expression protocol](https://www.10xgenomics.com/products/spatial-gene-expression) and has not been tested using output from other spatial transcriptomics methods._

> ## Output files

Author: > maybe we can have a section of "Workflow options" and throw a few (or at least one) other things in there?

I like this concept, but nothing is really jumping out at me as along belonging in that section. Though, one day I'm sure more things will arise! The only "maybe" is to also move the section on _repeating_ cell typing into it too?
